overlord/snapstate/handlers.go

```diff
@@ -3512,16 +3520,113 @@ func (m *SnapManager) doRemoveAliases(t *state.Task, _ *tomb.Tomb) error {
 	}
 	snapName := snapsup.InstanceName()
 
-	err = m.backend.RemoveSnapAliases(snapName)
-	if err != nil {
+	var removeReason removeAliasesReason
+	if err := t.Get("remove-reason", &removeReason); err != nil && !errors.Is(err, state.ErrNoState) {
+		return err
+	}
+
+	tr := config.NewTransaction(st)
+	experimentalRefreshAppAwareness, err := features.Flag(tr, features.RefreshAppAwareness)
+	if err != nil && !config.IsNoOption(err) {
 		return err
 	}
 
+	if removeReason != removeAliasesReasonRefresh || !experimentalRefreshAppAwareness || excludeFromRefreshAppAwareness(snapsup.Type) {
+		err = m.backend.RemoveSnapAliases(snapName)
+		if err != nil {
+			return err
+		}
+	} else {
+		t.Set("remove-aliases-skipped", true)
```

Comment:
do we need this or could we simply not set AliasesPending on this path?

Reply:
This is an interesting approach, I like it a lot. `setup-aliases` will then need to check if `remove-aliases` is skipped in an alternative way, maybe checking for the same conditions that caused `remove-aliases` to skip?

something like: `setupReason == "refresh" && experimentalRefreshAppAwareness && !excludeFromRefreshAppAwareness(snapsup.Type)`, but that brings an interesting question, could refresh-app-awareness flag be updated in the middle of a change? if yes, then I would need to find another reliable alternative to detect if `remove-aliases` was skipped.